### PR TITLE
Start of process manager

### PR DESCRIPTION
**Pull request by acron0**
Still need to:
* Replace batch-id with campaign id?
* Clean up PM database

### CODE REVIEW

File: src/kixi/collect/definitions.clj [hunk 1]
```diff
@@ -26,18 +43,25 @@
                 ::cr/groups
                 ::ms/id]))
 
+(defmethod comms/command-payload
```
Comment (acron0) [inline src/kixi/collect/definitions.clj]: Remove this

File: src/kixi/collect/process_manager/collection_request.clj [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+  ([backend event]
+   (advance-state-machine! backend event nil))
+  ([backend event start-state]
+   (let [old-state (pm/get-state backend event)]
+     (when (or old-state start-state)
+       (let [{:keys [value] :as new-state} (a/advance compiled-state-machine (or old-state start-state) event)
+             new-cmd-id (uuid)
+             action-value (::action value)
+             action-fn (get action-value->command-fn action-value)
+             action-result-cmd (when action-fn
```
Comment (acron0) [inline src/kixi/collect/process_manager/collection_request.clj]: Disallow returning multiple commands at this stage

File: src/kixi/collect/process_manager/collection_request.clj [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+      (mapv (fn [[group-id collection-request-id]]
+              (advance-state-machine!
+               backend
+               (assoc event ::cr/group-collection-requests {group-id collection-request-id})
+               (assoc start-state ::cr/id collection-request-id)))
+            group-collection-requests))))
+
+(defn handle-complete-process-command
+  [command]
+  [(merge {:kixi.event/type :kixi.collect.process-manager.collection-request/process-completed
```
Comment (acron0) [inline src/kixi/collect/process_manager/collection_request.clj]: Add another batch check?!

File: src/kixi/collect/process_manager/collection_request.clj [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+  (let [bid (get-in state [:value ::batch-id])
+        this-cr-id (get-in state [:value ::cr/id])
+        batch-results (get-batch backend bid)
+        requests-in-batch (get-in state [:value ::cr/ids])
+        grpd-batch-results (->> (update (group-by ::cr/id batch-results) this-cr-id conj {::action "complete"})
+                                (vals)
+                                (map #(set (map (comp keyword ::action) %))))]
+    (when (every? (partial = state-machine-actions) grpd-batch-results)
+      (log/info (str "Collection Request process was completed (batch " bid ", " (count batch-results)")"))
+      [(merge {:kixi.command/type :kixi.collect.process-manager.collection-request/complete-process
```
Comment (acron0) [inline src/kixi/collect/process_manager/collection_request.clj]: Add documentation re having events return events

Comment (tcoupland) [reply src/kixi/collect/process_manager/collection_request.clj]: events should not return events

File: test/kixi/integration/aggregate_test.clj [hunk 5]
```diff
@@ -5,40 +5,55 @@
...
+        dr (empty-datapack uid)
+        message :not-a-string
+        groups (random-uuid-set)]
+    (when-success dr
+      (stest/unstrument)
+      (binding [comms/*validate-commands* false]
+        (send-request-cmd uid message groups (get-in dr [:body ::ms/id])))
+      (let [event (wait-for-events uid :kixi.collect/collection-request-rejected)]
+        (is (= :invalid-cmd (:kixi.collect.request.rejection/reason event))))
+      (stest/instrument))))
```
Comment (otfrom) [inline test/kixi/integration/aggregate_test.clj]: do you want to remove this instrumentation?

**Review by otfrom — APPROVED** [review]

**Pull request merged**